new/rfcXXX-policyfile-endpoint.md

```diff
@@ -0,0 +1,485 @@
+
+### API Schema
+
+#### `/policy_groups`
+
+Container for policy groups.
+
+##### `GET /policy_groups`
+
+Returns a list of policy groups that exist.
+
+#### `/policy_groups/:policy_group_name/_acl`
+
+Authorization endpoint for the policy group `:policy_group_name`
+
+#### `/policy_groups/:policy_group_name/policies/`
+
+##### `GET /policy_groups/:policy_group_name/policies/`
+
+Returns a JSON object showing the active policy names and revision IDs for the
+policy group `:policy_group_name`.
+
+#### `/policy_groups/:policy_group_name/policies/:policy_name`
+
+##### `GET /policy_groups/:policy_group_name/policies/:policy_name`
+
+Returns the policyfile lock document at the revision that is currently
+associated with the given policy group and policy name.
+
+##### `PUT /policy_groups/:policy_group_name/policies/:policy_name`
+
+Sets the active revision of `:policy_name` for the policy group
+`:policy_group_name` to the policyfile lock document in the request body. The
+policy group, policy name, and policyfile lock revision will be created if any
+does not exist.
+
+##### `POST /policy_groups/:policy_group_name/policies/:policy_name`
+
+Sets the current active revision of `:policy_name` for the `:policy_group_name`
+to the value given in the POST body. The revision must already exist on the
+server. For example, to set the active revision of the 'appserver' policy in
+the 'qa' policy group, you would POST to
+`/policy_groups/qa/policies/appserver`, with a POST body like the following:
+
+```json
+{
+  "revision_id": "bfec256f76a52ff707ca72e71b464a41c410b229"
+}
+```
+
+#### `/policy_groups/:policy_group_name/nodes`
+
+##### `GET /policy_groups/:policy_group_name/nodes`
+
+Returns a list of nodes that belong to the policy group. Can be filtered for a
+specific `:policy_name` with the `policy_name` query parameter, using a URL of
+the form: `GET /policy_groups/:policy_group_name/nodes?policy_name=:policy_name`
+
+#### `/policies/:policy_name`
+
+TODO: This is where AuthZ for policies goes, does it do anything else?
+
+This implies that a PUT to /policy\_groups/foo/policies/bar has to authorize
+against both policies/\_acl and policy\_groups/\_acl which maybe is weird (?).
+
+#### `/policies/:policy_name/revisions/`
+
+##### `POST /policies/:policy_name/revisions/`
+
+Create a new Policyfile lock document. The POST body MUST be a valid Policyfile
+lock document, as described above. The document can subsequently be retrieved
+at the relative URL `/policies/:policy_name/revisions/:revision_id`
+
+##### `GET /policies/:policy_name/revisions/`
+
+Returns a list of the available revisions for the given policy name
+
+#### `/policies/:policy_name/revisions/:revision_id`
+
+##### `GET /policies/:policy_name/revisions/:revision_id`
+
+Returns the policyfile lock document with the given name and revision ID.
+
+##### `DELETE /policies/:policy_name/revisions/:revision_id`
+
+Deletes the policyfile lock document with the given name and revision ID.
```

Comment:
From @manderson26 via chat, this implies that a malicious/mistaken/drunken actor can delete a policy that's in-use in a group they don't have access to (see authz section). This should probably return 409 if the revision is active in any group. We maybe could offer a force option, which would require that you be authorized to update every group currently using the policy.